cilium/CFP-22215-mutual-auth-for-service-mesh.md

```diff
@@ -0,0 +1,442 @@
+1. Cilium starts up as normal, acts as CNI.
+1. Cilium also contacts the local SPIRE agent at startup (via a domain socket shared on the host's filesystem) to watch the Delegated Identity API and gets its own SPIFFE identity via the SPIRE workload API.
+1. When generating Cilium Security Identities for identities with mTLS auth enabled, Cilium (could be agent or operator) also records the SPIFFE identity.
+1. When checking auth-enabled policy, the datapath will flag that there is no entry in the auth table for the flow, and then the agent will use its cache of keypairs to handshakes with the agent on the other end of the connection (either source node agent will reach out to destination node agent, or vice versa), and performs a TLS handshake. Note that an optimization which may be included here is that the agent may pre-authenticate identities that it knows will need it, if we can make that work.
+1. If the TLS handshake succeeds, then the connection is authenticated and the traffic protected by policy can proceed. The TLS connection between the agents can then be torn down and thrown away (since it’s not being used to tunnel the encrypted traffic through).
+
+
+### Control Plane
+
+#### Structures
+
+Auth table - a BPF table needs to be added that records, at least:
+
+* Source Identity
+* Destination Identity
+* Destination Node
+* Authentication expiry time
+
+Further details of this table will be worked out as part of the implementation.
+
+#### Connections
+
+TODO: This section needs updating
+
+##### Between Cilium Operator instances and the SPIFFE server
+
+SPIFFE identity registration must occur out-of-band from when the actual SPIFFE identity usage is made via the workload API. In order to register the SPIFFE identities, the Cilium Operator must share access to an admin domain socket with the SPIRE Server, and use that domain socket for creation of new SPIFFE identities in the SPIRE server.
+
+Notably, this means that _one_ Pod out of the Operator deployment must be running on the _same node_ as the SPIRE Server's StatefulSet Pod. This will need Pod Affinity, and some way for the Operator to check if the domain socket is available. This process will also need to run independently of the usual Kubernetes Leadership, unfortunately.
+
+##### Between Cilium Agent instance peers
+Mutual authentication connections are established between the node IPs themselves. However, the peers that require authentication will not be these nodes.
+
+The source agent determines the destination agent using the destination endpoint IP. It then reaches out and connects to the listening port on the destination agent, and completes a TLS handshake. After the TLS handshake is complete, the session is torn down, as all the relevant information has been required. (Namely, did the authentication work, and when do the client and server expire?)
```

Comment:
this explains the egress case. we should also mention the ingress case where this works vice-versa.

Reply:
Doesn't have to be if the receiving note sends down that it got a connection, but this should be looked into and noted which approach to take (or both)

Reply:
I've updated this; I think that both nodes should record that the handshake was completed.